File: golem-worker-executor/tests/agent_sdk_go/promise.rs
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Golem Source License v1.1 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! The Go SDK's promise wrapper: awaiting a promise durably SUSPENDS the worker
//! (it does not spin or return early), and completing it from outside resumes the
//! agent with the payload. Mirrors `api::promise`.

use crate::Tracing;
use anyhow::anyhow;
use golem_common::model::{AgentStatus, OplogIndex, PromiseId};
use golem_common::{agent_id, data_value};
use golem_test_framework::dsl::TestDsl;
use golem_worker_executor_test_utils::{
    LastUniqueId, PrecompiledComponent, TestContext, WorkerExecutorTestDependencies, start,
};
use std::collections::HashMap;
use std::time::Duration;
use test_r::{inherit_test_dep, test, timeout};
use tracing::Instrument;

inherit_test_dep!(WorkerExecutorTestDependencies);
inherit_test_dep!(LastUniqueId);
inherit_test_dep!(Tracing);
inherit_test_dep!(
    #[tagged_as("agent_sdk_go")]
    PrecompiledComponent
);

/// The agent creates a promise and awaits it in a second invocation; the worker
/// must reach Suspended rather than returning, and completing the promise through
/// the executor API resumes it with the completed payload.
#[test]
#[tracing::instrument]
#[timeout("2m")]
async fn go_promise_suspends_until_completed(
    last_unique_id: &LastUniqueId,
    deps: &WorkerExecutorTestDependencies,
    _tracing: &Tracing,
    #[tagged_as("agent_sdk_go")] agent_sdk_go: &PrecompiledComponent,
) -> anyhow::Result<()> {
    let context = TestContext::new(last_unique_id);
    let executor = start(deps, &context).await?;

    let component = executor
        .component_dep(&context.default_environment_id, agent_sdk_go)
        .store()
        .await?;
    let agent_id = agent_id!("PromiseAgent", "go-promise-1");
    let worker_id = executor
        .start_agent_with(&component.id, agent_id.clone(), HashMap::new(), Vec::new())
        .await?;

    // The agent returns the promise's oplog index, which addresses it for completion.
    let oplog_idx = executor
        .invoke_and_await_agent(&component, &agent_id, "create", data_value!())
        .await?
        .into_typed::<i64>()?;

    let executor_clone = executor.clone();
    let component_clone = component.clone();
    let agent_id_clone = agent_id.clone();
    let mut fiber = tokio::spawn(
        async move {
            executor_clone
                .invoke_and_await_agent(
                    &component_clone,
                    &agent_id_clone,
                    "await",
                    data_value!(oplog_idx),
                )
                .await
        }
        .in_current_span(),
    );

    // The await must park the worker rather than complete.
    tokio::select! {
        result = &mut fiber => {
            let invoke_result = result??;
            return Err(anyhow!("await returned instead of suspending: {invoke_result:?}"));
        }
        status = executor.wait_for_status(&worker_id, AgentStatus::Suspended, Duration::from_secs(10)) => {
            status?;
        }
    }

    executor
        .complete_promise(
            &PromiseId {
                agent_id: worker_id.clone(),
                oplog_idx: OplogIndex::from_u64(oplog_idx as u64),
            },
            // The Go wrapper decodes the payload as JSON for a string T.
            b"\"approved\"".to_vec(),
        )
        .await?;

    let awaited = fiber.await??.into_typed::<String>()?;

    executor.check_oplog_is_queryable(&worker_id).await?;
    drop(executor);

    assert_eq!(awaited, "approved");
    Ok(())
}

File: sdks/go/golem/promise.go
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package golem

import (
	"encoding/json"
	"fmt"

	host "github.com/golemcloud/golem/sdks/go/golem/internal/wit/golem_agent_host"
	apiHost "github.com/golemcloud/golem/sdks/go/golem/internal/wit/golem_api_host"
	types "github.com/golemcloud/golem/sdks/go/golem/internal/wit/golem_core_types"
)

// Promises are Golem's durable primitive for human-in-the-loop and asynchronous
// external integration. An agent creates a promise, hands its [PromiseID] to the
// outside world, and awaits it; the invocation durably SUSPENDS — surviving
// restarts, exactly-once via oplog replay — until someone completes it with a
// payload.
//
// Completion can come from another agent ([CompletePromise]) or from off-platform
// via the worker REST API (POST /:component/workers/:agent/complete with the
// PromiseID's fields), so the completion payload is arbitrary bytes with no
// predefined meaning. [Promise] types those bytes as T via JSON, so an external
// completer (a webhook, a curl, a human) can produce them. As an escape hatch, a
// Promise[[]byte] carries the raw bytes through unchanged.

// PromiseID is the durable, serializable identity of a promise. Hand it to
// whoever will complete the promise — another agent, or an external system via
// the worker-service complete endpoint (which needs exactly these three fields).
// The exported fields make it directly JSON-marshalable.
type PromiseID struct {
	// ComponentID is the component the owning agent belongs to.
	ComponentID UUID
	// AgentID is the string agent id (agent type + constructor parameters) of the
	// agent that created the promise — the ":agent" path segment of the REST
	// complete endpoint.
	AgentID string
	// OplogIndex disambiguates promises created by the same agent — the "oplogIdx"
	// field of the REST complete endpoint.
	OplogIndex uint64
}

// String renders a compact, human-readable form (not a parse target).
func (id PromiseID) String() string {
	return fmt.Sprintf("%s/%s/%d", id.ComponentID, id.AgentID, id.OplogIndex)
}

func promiseIDFromWit(w types.PromiseId) PromiseID {
	return PromiseID{
		ComponentID: uuidFromWit(w.AgentId.ComponentId.Uuid),
		AgentID:     w.AgentId.AgentId,
		OplogIndex:  w.OplogIdx,
	}
}

func (id PromiseID) toWit() types.PromiseId {
	return types.PromiseId{
		AgentId: types.AgentId{
			ComponentId: types.ComponentId{Uuid: uuidToWit(id.ComponentID)},
			AgentId:     id.AgentID,
		},
		OplogIdx: id.OplogIndex,
	}
}

// Promise is a handle to a durable promise whose completion payload is T. Obtain
// one with [NewPromise] inside an agent, or rebuild it from a stored id with
// [PromiseByID]. Only the agent that created the promise may [Promise.Await] it.
type Promise[T any] struct {
	id PromiseID
}

// NewPromise creates a fresh promise owned by the current agent and returns a
// handle to it. Call it from inside an agent (a method or the constructor); the
// host mints the id. Persist [Promise.ID] in the agent's state (or send it out)
// if the promise will be awaited or completed in a later invocation.
func NewPromise[T any]() *Promise[T] {
	return &Promise[T]{id: promiseIDFromWit(apiHost.CreatePromise())}
}

// PromiseByID rebuilds a handle to an existing promise from a stored [PromiseID]
// — e.g. an id saved in agent state during an earlier invocation, so a later
// invocation of the same agent can [Promise.Await] it.
func PromiseByID[T any](id PromiseID) *Promise[T] {
	return &Promise[T]{id: id}
}

// ID returns the durable identity of the promise, to persist or hand to a
// completer.
func (p *Promise[T]) ID() PromiseID { return p.id }

// Await blocks until the promise is completed and returns the payload decoded as
// T (JSON, or the raw bytes when T is []byte). It durably SUSPENDS the invocation
// — the worker may be evicted and resumed, and on replay the recorded payload is
// returned from the oplog (exactly-once). Like [Future.Get] it is fail-loud: an
// infra failure traps and surfaces as an agent-error rather than returning an
// error value.
//
// Only the agent that created the promise may Await it; awaiting from another
// agent traps in the host.
func (p *Promise[T]) Await() T {
	res := apiHost.GetPromise(p.id.toWit())
	data := res.Get()
	res.Drop()
	return decodePromisePayload[T](data)
}

// WebhookURL mints an external URL that completes this promise: a POST to the URL
// completes the promise with the request body, so a later [Promise.Await] returns
// that body decoded as T (JSON, or the raw bytes when T is []byte). Hand the URL
// to an off-platform system to drive a webhook-style callback.
//
// It requires the agent type to be currently deployed behind an HTTP API (set
// [Spec].HTTP on the agent and declare the agent in the manifest's httpApi
// deployment) — it is fail-loud: the runtime traps, surfacing an agent-error, if
// the agent is not http-api-deployed, or if a different agent type created the
// promise. Typically called once per promise.
func (p *Promise[T]) WebhookURL() string {
	res := host.CreateWebhook(p.id.toWit())
	if res.IsErr() {
		panic(webhookErrorToGo(res.Err()))
	}
	return res.Ok()
}

func webhookErrorToGo(e host.WebhookError) error {
	switch e.Tag() {
	case host.WebhookErrorPermissionDenied:
		return fmt.Errorf("golem: webhook: permission denied")
	case host.WebhookErrorInternalError:
		return fmt.Errorf("golem: webhook: internal error: %s", e.InternalError())
	default:
		return fmt.Errorf("golem: webhook: error (tag %d)", e.Tag())
	}
}

// CompletePromise completes the promise identified by id, from within any agent
// (agent-to-agent completion; external completers use the REST endpoint instead).
// The value is encoded as the payload (JSON, or raw bytes when T is []byte).
//
// It returns true if this call completed the promise, false if it was already
// completed — the idempotency signal from the host. It is fail-loud only on an
// infra/encode failure.
func CompletePromise[T any](id PromiseID, value T) bool {
	return apiHost.CompletePromise(id.toWit(), encodePromisePayload(value))
}

// encodePromisePayload serializes a promise payload: raw pass-through when T is
// exactly []byte (so non-JSON external payloads round-trip), otherwise JSON.
// Pure (no host import), so it is natively testable.
func encodePromisePayload[T any](v T) []byte {
	if b, ok := any(v).([]byte); ok {
		return b
	}
	b, err := json.Marshal(v)
	if err != nil {
		panic(fmt.Errorf("golem: encoding promise payload: %w", err))
	}
	return b
}

// decodePromisePayload is the inverse of [encodePromisePayload].
func decodePromisePayload[T any](data []byte) T {
	var out T
	if _, ok := any(out).([]byte); ok {
		return any(data).(T)
	}
	if err := json.Unmarshal(data, &out); err != nil {
		panic(fmt.Errorf("golem: decoding promise payload: %w", err))
	}
	return out
}
